src/Symfony/Component/Routing/Generator/UrlGenerator.php

```diff
@@ -78,30 +95,124 @@ class UrlGenerator implements UrlGeneratorInterface, ConfigurableRequirementsInt
+     * be used when generating an absolute URI for a route without specific scheme or host requirement. The default base URI
+     * before calling this method is "http://localhost/".
+     *
+     * @param string|UriInterface $uri The base URI as string or PSR-7 UriInterface implementation
+     */
+    public function setBaseUri($uri)
+    {
+        if ($uri instanceof UriInterface) {
+            $this->scheme = $uri->getScheme() ?: $this->scheme;
+            $this->host = $uri->getHost() ?: $this->host;
+            $this->path = $uri->getPath() ?: $this->path;
+            $port = $uri->getPort();
+        } else {
+            $uriComponents = parse_url($uri);
+
+            if (false === $uriComponents) {
+                throw new \InvalidArgumentException('The base URI string appears to be malformed.');
+            }
+
+            $this->scheme = isset($uriComponents['scheme']) ? strtolower($uriComponents['scheme']) : $this->scheme;
+            $this->host = isset($uriComponents['host']) ? strtolower($uriComponents['host']) : $this->host;
+            $this->path = isset($uriComponents['path']) ? $uriComponents['path'] : $this->path;
+            $port = isset($uriComponents['port']) ? $uriComponents['port'] : null;
+        }
+
+        if (null !== $port) {
+            if ('http' === $this->scheme) {
+                $this->httpPort = $port;
+            } elseif ('https' === $this->scheme) {
+                $this->httpsPort = $port;
+            }
+        }
+    }
+
+    /**
+     * Sets the script path used as path prefix for generated URIs.
+     *
+     * This is usually the front controller path, e.g. "/app_dev.php", that all paths should start with. It should be empty,
+     * which is the default, when URI rewriting is used or the URI should not contain the script. The naming is based on
+     * SCRIPT_NAME in the CGI spec which is also available in the $_SERVER global in PHP.
+     *
+     * @param string $scriptName The URL encoded script name
+     */
+    public function setScriptName($scriptName)
+    {
+        $this->scriptName = $scriptName;
+    }
+
+    /**
+     * Sets a parameter value that will be used for placeholders by default.
+     *
+     * This placeholder parameter will be used if none has not been provided explicitly in the generate() method.
+     *
+     * @param string           $name  A parameter name
+     * @param string|int|float $value The parameter value
+     */
+    public function setDefaultParameter($name, $value)
+    {
+        $this->defaultParameters[$name] = $value;
     }
 
     /**
      * {@inheritdoc}
+     *
+     * @deprecated since version 2.8, to be removed in 3.0. Use setBaseUri and setScriptName instead.
      */
     public function setContext(RequestContext $context)
     {
+        @trigger_error('The '.__METHOD__.' method is deprecated since version 2.8 and will be removed in 3.0. Use setBaseUri and setScriptName instead.', E_USER_DEPRECATED);
```

Comment:
IMO, the message should be built with `sprintf()`, by instance:
```php
@trigger_error(sprintf('The "%s()" method is deprecated since version 2.8 and will be removed in 3.0. Use "setBaseUri()" and "setScriptName()" instead.', __METHOD__), E_USER_DEPRECATED);
```